### PR TITLE
implement getNodeInfo features

### PR DESCRIPTION
**Pull request by howjmay**
Use bitwise enum `node_info_mask_feature_e` to records which functionality
is able to be use on this machine.

`features_set()` can set the `features` field of `get_node_info_res_t`

associate #1031

### CODE REVIEW

File: cclient/response/get_node_info.c [hunk 1]
```diff
@@ -16,6 +16,8 @@ get_node_info_res_t* get_node_info_res_new() {
     memset(res->latest_milestone, FLEX_TRIT_NULL_VALUE, FLEX_TRIT_SIZE_243);
     memset(res->latest_solid_subtangle_milestone, FLEX_TRIT_NULL_VALUE, FLEX_TRIT_SIZE_243);
     memset(res->coordinator_address, FLEX_TRIT_NULL_VALUE, FLEX_TRIT_SIZE_243);
+    res->features = NULL;
+    utarray_new(res->features, &ut_str_icd);
```
Comment (thibault-martinez) [inline cclient/response/get_node_info.c:19]: No need to new it here since you do it in `get_node_info_res_features_add`

File: gossip/feature.h [hunk 2]
```diff
@@ -0,0 +1,31 @@
+#include <stdint.h>
+#include <stdlib.h>
+#include "utarray.h"
+
+#define FEATURE_TYPES_NUM 6
```
Comment (thibault-martinez) [inline gossip/feature.h]: Just call it `FEATURES_NUM`

File: gossip/node.h [hunk 3]
```diff
@@ -42,6 +43,7 @@ typedef struct node_s {
   neighbor_t* neighbors;
   rw_lock_handle_t neighbors_lock;
   tips_cache_t tips;
+  uint32_t features;
```
Comment (thibault-martinez) [inline gossip/node.h]: An `uint8_t` is currently sufficient, we'll increase it later if needed.

File: gossip/feature.c [hunk 4]
```diff
@@ -0,0 +1,12 @@
+#include "feature.h"
+
+void features_set(UT_array* feature_ut, uint32_t features) {
+  node_info_mask_feature_e elt = (1u << 0);
```
Comment (thibault-martinez) [inline gossip/feature.c]: new line

File: gossip/feature.c [hunk 5]
```diff
@@ -0,0 +1,12 @@
+#include "feature.h"
+
+void features_set(UT_array* feature_ut, uint32_t features) {
```
Comment (thibault-martinez) [inline gossip/feature.c]: Can you reverse parameters ?

File: gossip/feature.c [hunk 6]
```diff
@@ -0,0 +1,12 @@
+#include "feature.h"
```
Comment (thibault-martinez) [inline gossip/feature.c]: Copyright

File: gossip/feature.h [hunk 7]
```diff
@@ -0,0 +1,31 @@
+#include <stdint.h>
```
Comment (thibault-martinez) [inline gossip/feature.h]: Copyright

File: gossip/feature.h [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+#include <stdlib.h>
+#include "utarray.h"
+
+#define FEATURE_TYPES_NUM 6
+
+char* const feature_str[FEATURE_TYPES_NUM] = {"snapshotPruning",  "dnsRefresher",      "testnet",
+                                              "zeroMessageQueue", "tipSolidification", "RemotePOW"};
+
+typedef enum node_info_mask_feature {
```
Comment (thibault-martinez) [inline gossip/feature.h]: `typedef enum node_info_mask_feature_e`

File: gossip/feature.h [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+  MASK_FEATURE_SNAPSHOT_PRUNING = (1u << 0),
+  MASK_FEATURE_DNS_REFRESHER = (1u << 1),
+  MASK_FEATURE_TESTNET = (1u << 2),
+  MASK_FEATURE_ZERO_MESSAGE_QUEUE = (1u << 3),
+  MASK_FEATURE_TIP_SOLIDIFICATION = (1u << 4),
+  MASK_FEATURE_REMOTE_POW = (1u << 5),
+  MASK_FEATURE_ALL = MASK_FEATURE_SNAPSHOT_PRUNING | MASK_FEATURE_DNS_REFRESHER | MASK_FEATURE_TESTNET |
+                     MASK_FEATURE_ZERO_MESSAGE_QUEUE | MASK_FEATURE_TIP_SOLIDIFICATION | MASK_FEATURE_REMOTE_POW
+} node_info_mask_feature_e;
```
Comment (thibault-martinez) [inline gossip/feature.h]: `node_info_mask_feature_t`

**Review by thibault-martinez — CHANGES_REQUESTED** [review]
Hi @HowJMay ! Thank you very much for this really great PR !
Some quick nits to begin :)

File: gossip/feature.h [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+  *node_features |= input_feature;
+}
+
+static inline void get_node_info_feature_clear(uint32_t* node_features, node_info_mask_feature_t input_feature) {
+  *node_features &= (!input_feature);
+}
+
+static inline void get_node_info_feature_clear_all(uint32_t* node_features) { *node_features = 0; }
+
+void features_set(uint32_t features, UT_array* feature_ut);
```
Comment (thibault-martinez) [inline gossip/feature.h]: `uint8_t`

File: gossip/feature.h [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+
+static inline void get_node_info_feature_set(uint32_t* node_features, node_info_mask_feature_t input_feature) {
+  *node_features |= input_feature;
+}
+
+static inline void get_node_info_feature_clear(uint32_t* node_features, node_info_mask_feature_t input_feature) {
+  *node_features &= (!input_feature);
+}
+
+static inline void get_node_info_feature_clear_all(uint32_t* node_features) { *node_features = 0; }
```
Comment (thibault-martinez) [inline gossip/feature.h]: `uint8_t`

File: gossip/feature.h [hunk 12]
```diff
@@ -0,0 +1,38 @@
...
+  MASK_FEATURE_REMOTE_POW = (1u << 5),
+  MASK_FEATURE_ALL = MASK_FEATURE_SNAPSHOT_PRUNING | MASK_FEATURE_DNS_REFRESHER | MASK_FEATURE_TESTNET |
+                     MASK_FEATURE_ZERO_MESSAGE_QUEUE | MASK_FEATURE_TIP_SOLIDIFICATION | MASK_FEATURE_REMOTE_POW
+} node_info_mask_feature_t;
+
+static inline void get_node_info_feature_set(uint32_t* node_features, node_info_mask_feature_t input_feature) {
+  *node_features |= input_feature;
+}
+
+static inline void get_node_info_feature_clear(uint32_t* node_features, node_info_mask_feature_t input_feature) {
```
Comment (thibault-martinez) [inline gossip/feature.h]: `uint8_t`

File: gossip/feature.h [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+  MASK_FEATURE_DNS_REFRESHER = (1u << 1),
+  MASK_FEATURE_TESTNET = (1u << 2),
+  MASK_FEATURE_ZERO_MESSAGE_QUEUE = (1u << 3),
+  MASK_FEATURE_TIP_SOLIDIFICATION = (1u << 4),
+  MASK_FEATURE_REMOTE_POW = (1u << 5),
+  MASK_FEATURE_ALL = MASK_FEATURE_SNAPSHOT_PRUNING | MASK_FEATURE_DNS_REFRESHER | MASK_FEATURE_TESTNET |
+                     MASK_FEATURE_ZERO_MESSAGE_QUEUE | MASK_FEATURE_TIP_SOLIDIFICATION | MASK_FEATURE_REMOTE_POW
+} node_info_mask_feature_t;
+
+static inline void get_node_info_feature_set(uint32_t* node_features, node_info_mask_feature_t input_feature) {
```
Comment (thibault-martinez) [inline gossip/feature.h]: `uint8_t`

File: gossip/feature.h [hunk 14]
```diff
@@ -0,0 +1,38 @@
+/*
+ * Copyright (c) 2019 IOTA Stiftung
+ * https://github.com/iotaledger/entangled
+ *
+ * Refer to the LICENSE file for licensing information
+ */
+
+#include <stdint.h>
+#include <stdlib.h>
```
Comment (thibault-martinez) [inline gossip/feature.h]: new line

File: gossip/feature.c [hunk 15]
```diff
@@ -0,0 +1,20 @@
+/*
+ * Copyright (c) 2019 IOTA Stiftung
+ * https://github.com/iotaledger/entangled
+ *
+ * Refer to the LICENSE file for licensing information
+ */
+
+#include "feature.h"
+
+void features_set(uint32_t features, UT_array* feature_ut) {
```
Comment (thibault-martinez) [inline gossip/feature.c]: `uint8_t`

**Review by thibault-martinez — CHANGES_REQUESTED** [review]

File: cclient/response/get_node_info.h [hunk 16]
```diff
@@ -67,6 +67,10 @@ typedef struct get_node_info_res_s {
    * of all referenced transactions.
    */
   flex_trit_t latest_solid_subtangle_milestone[FLEX_TRIT_SIZE_243];
+  /*
+   * Enabled commands
```
Comment (thibault-martinez) [inline cclient/response/get_node_info.h]:
```suggestion
   * Enabled features
```

File: ciri/api/api.c [hunk 17]
```diff
@@ -28,7 +28,26 @@ static bool invalid_subtangle_status(iota_api_t const *const api, error_res_t **
...
+  res->neighbors = neighbors_count(api->core->node.neighbors);
+  rw_lock_handle_unlock(&api->core->node.neighbors_lock);
+  res->packets_queue_size = broadcaster_size(&api->core->node.broadcaster);
+  res->time = current_timestamp_ms();
+  res->tips = tips_cache_size(&api->core->node.tips);
+  res->transactions_to_request = requester_size(&api->core->node.transaction_requester);
+  features_set(res->features, &api->core->node.features);
+  memcpy(res->coordinator_address, api->core->consensus.conf.coordinator_address, FLEX_TRIT_SIZE_243);
+
+  return RC_OK;
```
Comment (thibault-martinez) [inline ciri/api/api.c]: I think there is a failed conflict merging here, you are modifying the function `invalid_subtangle_status`

File: gossip/node.h [hunk 18]
```diff
@@ -42,6 +43,7 @@ typedef struct node_s {
   neighbor_t* neighbors;
   rw_lock_handle_t neighbors_lock;
   tips_cache_t tips;
+  uint8_t features;
```
Comment (thibault-martinez) [inline gossip/node.h]: Can you move this to the API ? Some of the features are not node-related.

**Review by thibault-martinez — CHANGES_REQUESTED** [review]

File: gossip/node.h (hunk 18 above)
Comment (howjmay) [reply gossip/node.h]: so I should move this to `ciri/api/api.h`?

Comment (thibault-martinez) [reply gossip/node.h]: Yes please

**Comment by thibault-martinez** [comment]
@HowJMay Please try to locally run tests when you're done, this is currently not compiling.

**Comment by howjmay** [comment]
Sorry for that 
I will run it as soon as I get home

**Comment by thibault-martinez** [comment]
@HowJMay Seems like there is a leak, can you check please ?

**Review by thibault-martinez — APPROVED** [review]
Thanks @HowJMay !

**Pull request merged**